mempool/priority/mempool.go

```diff
@@ -0,0 +1,809 @@
+func (txmp *TxMempool) canAddTx(wtx *WrappedTx) error {
+	numTxs := txmp.Size()
+	txBytes := txmp.SizeBytes()
+
+	if numTxs >= txmp.config.Size || wtx.Size()+txBytes > txmp.config.MaxTxsBytes {
+		return mempool.ErrMempoolIsFull{
+			NumTxs:      numTxs,
+			MaxTxs:      txmp.config.Size,
+			TxsBytes:    txBytes,
+			MaxTxsBytes: txmp.config.MaxTxsBytes,
+		}
+	}
+
+	return nil
+}
+
+// CheckToPurgeExpiredTxs checks if there has been adequate time since the last time
+// the txpool looped through all transactions and if so, performs a purge of any transaction
+// that has expired according to the TTLDuration. This is thread safe.
+func (txmp *TxMempool) CheckToPurgeExpiredTxs() {
+	txmp.mtx.Lock()
+	defer txmp.mtx.Unlock()
+	if txmp.config.TTLDuration > 0 && time.Since(txmp.lastPurgeTime) > txmp.config.TTLDuration {
+		txmp.purgeExpiredTxs(txmp.height)
+	}
+}
+
+// purgeExpiredTxs removes all transactions from the mempool that have exceeded
+// their respective height or time-based limits as of the given blockHeight.
+// Transactions removed by this operation are not removed from the cache.
+//
+// The caller must hold txmp.mtx exclusively.
+func (txmp *TxMempool) purgeExpiredTxs(blockHeight int64) {
+	if txmp.config.TTLNumBlocks == 0 && txmp.config.TTLDuration == 0 {
+		return // nothing to do
+	}
+
+	now := time.Now()
+	cur := txmp.txs.Front()
+	for cur != nil {
+		// N.B. Grab the next element first, since if we remove cur its successor
+		// will be invalidated.
+		next := cur.Next()
+
+		w := cur.Value.(*WrappedTx)
+		if txmp.config.TTLNumBlocks > 0 && (blockHeight-w.height) > txmp.config.TTLNumBlocks ||
+			txmp.config.TTLDuration > 0 && now.Sub(w.timestamp) > txmp.config.TTLDuration {
+			txmp.removeTxByElement(cur)
+			txmp.cache.Remove(w.tx)
+			txmp.evictedTxs.Push(w.tx)
+			txmp.metrics.ExpiredTxs.Add(1)
+		}
+		cur = next
+	}
+
+	txmp.lastPurgeTime = now
+}
+
+func (txmp *TxMempool) notifyTxsAvailable() {
+	if txmp.Size() == 0 {
+		return // nothing to do
+	}
+
+	if txmp.txsAvailable != nil && !txmp.notifiedTxsAvailable {
+		// channel cap is 1, so this will send once
+		txmp.notifiedTxsAvailable = true
+
+		select {
+		case txmp.txsAvailable <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (txmp *TxMempool) preCheck(tx types.Tx) error {
+	txmp.mtx.Lock()
+	defer txmp.mtx.Unlock()
+	if txmp.preCheckFn != nil {
+		return txmp.preCheckFn(tx)
+	}
+	return nil
+}
+
+// getTxPriority returns a naive tx priority based on the amount of the smallest denomination of the gas price
+// provided in a transaction.
+// NOTE: This implementation should be used with a great consideration as it opens potential attack vectors
+// where txs with multiple coins could not be prioritize as expected.
+func getTxPriority(fee, gas int64) int64 {
+	var priority int64
+
+	p := int64(math.MaxInt64)
+	gasPrice := fee / gas
+	if gasPrice > 0 {
+		p = gasPrice
+	}
+	if priority == 0 || p < priority {
+		priority = p
+	}
+
+	return priority
+}
```

Comment:
I don't get this algorithm? 

```
if priority == 0 || p < priority {
	priority = p
}
```
isn't priority always 0 in this case so this if statement always triggers, why have it?

In the past we've had problems with truncating with ints, so we usually multiply the fee by something like 1,000,000 to ensure that when the fees are lower than the gas they're not all 0's in the mempool and indistinguishable. (just need to watch for overflow)

Reply:
its the same as it was in the sdk. but we are switching the logic to calculate up there and send us only the priority